include/rx/ranges.hpp

```diff
@@ -1354,62 +1625,116 @@ struct foldl {
 
     template <class InputRange>
     constexpr T operator()(InputRange&& input) {
+        // Copying for reentrancy.
         auto copy = as_input_range(std::forward<InputRange>(input));
         RX_TYPE_ASSERT(is_finite<decltype(copy)>);
         T accum = init;
-        while (!copy.at_end()) {
+        if (RX_UNLIKELY(copy.at_end()))
+            return accum;
+        do {
             accum = func(accum, copy.get());
             copy.next();
-        }
+        } while (RX_LIKELY(!copy.at_end()));
         return accum;
     }
 };
 template <class T, class F>
-foldl(T&&, F &&)->foldl<RX_REMOVE_CVREF_T<T>, RX_REMOVE_CVREF_T<F>>;
+foldl(T&&, F &&)->foldl<remove_cvref_t<T>, remove_cvref_t<F>>;
 
+/*!
+    @brief Sum the input range.
+
+    The output type of the input range is required to implement `operator+`.
+*/
 struct sum {
-    constexpr sum() noexcept {}
+    constexpr sum() = default;
     template <class R>
-    [[nodiscard]] constexpr auto operator()(R&& input) noexcept {
-        using type = get_output_type_of_t<R>;
-        return std::forward<R>(input)
-               | foldl(
-                   type{}, [](type accum, auto&& x) constexpr { return accum + x; });
+    [[nodiscard]] constexpr auto operator()(R&& input) const noexcept {
+        using type = remove_cvref_t<get_output_type_of_t<R>>;
+        // Note: foldl is reentrant.
+        auto folder = foldl(
+            type{}, [](type accum, auto&& x) constexpr { return accum + x; });
+        return std::move(folder)(std::forward<R>(input));
     }
 };
 
-struct max {
-    constexpr max() noexcept {}
+/*!
+    @brief Produce the max value of the input range as an optional.
+
+    If the input is empty, "none" is returned.
+
+    The output type of the input range is compared using an instance of Compare, which is
+    `std::less<void>` by default.
+
+    @tparam Compare The comparison function.
+*/
+template <class Compare = std::less<void>>
+struct max : private Compare {
+    template <class... Args>
+    constexpr max(Args&&... args) noexcept : Compare(std::forward<Args>(args)...) {}
```

Comment:
That overload prevents the copy/move constructor to be called.

Reply:
You're right. Maybe it would be simpler to just accept an instance of `Compare` by value. Do you think that would be acceptable?

Reply:
I think yes

Reply:
That's also what the standard does in std::set and other